### PR TITLE
ARROW-12946: [C++] String swap case kernel

### PR DESCRIPTION
**Pull request by Christian8491**
This PR adds `swapcase` compute kernel for string.  It is similar to  `Python str.swapcase()`

### CODE REVIEW

**Comment by edponce** [comment]
LGTM. Need to add implementation for UTF8.

File: cpp/src/arrow/compute/kernels/scalar_string.cc [hunk 1]
```diff
@@ -443,6 +488,17 @@ struct AsciiLower {
...
+  std::transform(input, input + length, output, ascii_swapcase);
+}
+
+template <typename Type>
+struct AsciiSwapCase {
+  static Status Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    return StringDataTransform<Type>(ctx, batch, TransformAsciiSwapCase, out);
+  }
+};
+
```
Comment (edponce) [inline cpp/src/arrow/compute/kernels/scalar_string.cc:501]: Add `Utf8SwapCase`

**Comment by edponce** [comment]
Also, add Python binding to `swapcase`.

File: cpp/src/arrow/compute/kernels/scalar_string_test.cc [hunk 2]
```diff
@@ -395,6 +395,12 @@ TYPED_TEST(TestStringKernels, AsciiLower) {
                    "[\"aaazzæÆ&\", null, \"\", \"bbb\"]");
 }
 
+TYPED_TEST(TestStringKernels, AsciiSwapCase) {
+  this->CheckUnary("ascii_swapcase", "[]", this->type(), "[]");
+  this->CheckUnary("ascii_swapcase", "[\"aAazZæÆ&\", null, \"\", \"BbB\"]", this->type(),
+                   "[\"AaAZzæÆ&\", null, \"\", \"bBb\"]");
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_string_test.cc:401]: Can you add characters that don't have a casing here?

File: cpp/src/arrow/compute/kernels/scalar_string_test.cc [hunk 3]
```diff
@@ -493,6 +499,23 @@ TYPED_TEST(TestStringKernels, Utf8Lower) {
                                   CallFunction("utf8_lower", {invalid_input}));
 }
 
+TYPED_TEST(TestStringKernels, Utf8SwapCase) {
+  this->CheckUnary("utf8_swapcase", "[\"aAazZæÆ&\", null, \"\", \"b\"]", this->type(),
+                   "[\"AaAZzÆæ&\", null, \"\", \"B\"]");
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/scalar_string_test.cc:504]: Same here: add characters that don't have a casing.

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks for doing this! Just a couple comments.

File: cpp/src/arrow/compute/kernels/scalar_string.cc (hunk 1 above)
Comment (Christian8491) [reply cpp/src/arrow/compute/kernels/scalar_string.cc:531]: Addressed in [ba66c91](https://github.com/apache/arrow/pull/10855/commits/ba66c919e0f4e593fa0b70b193af79ea7ec76fa4)

File: cpp/src/arrow/compute/kernels/scalar_string_test.cc (hunk 2 above)
Comment (Christian8491) [reply cpp/src/arrow/compute/kernels/scalar_string_test.cc:401]: Sure. Addressed in [8f65c70](https://github.com/apache/arrow/pull/10855/commits/8f65c700ae86d03d594dcf8957f7f17a583b2a97)

File: cpp/src/arrow/compute/kernels/scalar_string_test.cc (hunk 3 above)
Comment (Christian8491) [reply cpp/src/arrow/compute/kernels/scalar_string_test.cc:506]: Also addressed in [8f65c70](https://github.com/apache/arrow/pull/10855/commits/8f65c700ae86d03d594dcf8957f7f17a583b2a97)

**Comment by Christian8491** [comment]
@lidavidm please enable CI

File: cpp/src/arrow/compute/kernels/scalar_string.cc [hunk 4]
```diff
@@ -318,6 +373,26 @@ struct UTF8LowerTransform : public CaseMappingTransform {
...
+    if (codepoint <= kMaxCodepointLookup) {
+      return lut_swapcase_codepoint[codepoint];
+    } else {
+      if (utf8proc_islower(codepoint)) {
+        return utf8proc_toupper(codepoint);
+      } else if (utf8proc_isupper(codepoint)) {
+        return utf8proc_tolower(codepoint);
+      } else {
+        return codepoint;
+      }
```
Comment (edponce) [inline cpp/src/arrow/compute/kernels/scalar_string.cc:387]: The function does not ends with a `return` statement. Move the last else-case `return codepoint` as the last statement of the function.

**Comment by edponce** [comment]
An alternative solution is to use helper functions already available, refer to https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/scalar_string.cc#L1391-L1408

Related notes in Arrow w.r.t. to lower/upper casing of UTF8 can be found in:
* https://github.com/apache/arrow/blob/master/python/pyarrow/tests/test_compute.py#L668-L693
* https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/scalar_string_test.cc#L578-L601

**Comment by nealrichardson** [comment]
> Meanwhile, RTools35 is using utf8proc 2.4; it builds utf8proc from source due to some other issue, but I believe it's using the system headers anyways looking at the compiler command since the system headers come before the self-built utf8proc ones.

Yeah the logs say 

```
-- Could NOT find utf8proc (missing: utf8proc_LIB) (found suitable version "2.4.0", minimum required is "2.2.0")
```

which doesn't make sense.

**Comment by Christian8491** [comment]
As @edponce suggested, I replaced some `utf8proc` functions with helper functions. After all the CI passes. This PR is ready for review.

**Review by pitrou — APPROVED** [review]
+1, thank you @Christian8491

**Comment by edponce** [comment]
Some tests invoke the incorrect kernel (ASCII test uses `utf8_capitalize` and/or viceversa). These "typos" are fixed in https://github.com/apache/arrow/pull/10869

**Pull request closed without merging**